util/autogen_testutils.py

```
@@ -0,0 +1,88 @@
+
+    # Create list of IPs to generate shared testutils code for. This is all IPs
+    # that have a DIF library, that the testutils functions can use.
+    ips_with_difs = []
+    for autogen_dif_filename in glob.iglob(str(autogen_dif_directory / "*.h")):
+        # NOTE: the line below takes as input a file path
+        # (/path/to/dif_uart_autogen.c) and returns the IP name in lower
+        # case snake mode (i.e., uart).
+        ip_name_snake = Path(autogen_dif_filename).stem[4:-8]
+        # NOTE: ip.name_long_* not needed for auto-generated files which
+        # are the only files (re-)generated in batch mode.
+        ips_with_difs.append(Ip(ip_name_snake, "AUTOGEN"))
+    ips_with_difs.sort(key=lambda ip: ip.name_snake)
+
+    # Create output directories if needed.
+    autogen_testutils_dir.mkdir(exist_ok=True)
+
+    # Auto-generate testutils files.
+    for suffix in [".h", ".c"]:
+        for testutils_template_path_str in glob.iglob(
+                str(testutils_templates_dir / f"*{suffix}.tpl")):
+            testutils_template_path = Path(testutils_template_path_str)
+
+            # Read in template, render it, and write it to the output file.
+            testutils_template = Template(testutils_template_path.read_text())
+            testutils = Path(autogen_testutils_dir /
+                             testutils_template_path.stem)
+            testutils.write_text(
+                testutils_template.render(
+                    ips_with_difs=ips_with_difs,
+                    autogen_banner=get_autogen_banner(
+                        "util/autogen_testutils.py", comment="//"),
+                    license_banner=get_license_banner(comment="//")))
+
+            # Format autogenerated file with clang-format.
+            assert (shutil.which("clang-format") and
+                    "ERROR: clang-format must be installed to format "
+                    " autogenerated code to pass OpenTitan CI checks.")
```

This could be done prior to the loop and fail early.

---

Fixed.